File: src/modules/auth/session.service.ts
import { randomUUID } from 'node:crypto';
import { env } from '../../config/env.js';
import { generateSecureToken, sha256 } from '../../shared/utils/crypto.js';
import { expiresIn } from '../../shared/utils/ttl.js';
import type { Role } from '../../types/roles.js';
import { User } from '../user/user.model.js';
import {
  RefreshSession,
  SESSION_REVOCATION_REASONS,
  type SessionRevocationReason,
} from './refresh-session.model.js';

export interface SessionMetadata {
  ip: string | null;
  userAgent: string | null;
}

export interface CreatedSession {
  sessionId: string;
  refreshToken: string;
}

export interface SessionPublic {
  id: string;
  ip: string | null;
  userAgent: string | null;
  createdAt: string;
  lastUsedAt: string | null;
  expiresAt: string;
  active: boolean;
  current: boolean;
}

// Refresh tokens are opaque 256-bit random values, never JWTs. A token carries
// no identity itself; it is just a key into a server-side session row. This is
// what makes per-session revocation and reuse detection possible. Only the
// SHA-256 hash of a token is stored, so a leaked database cannot be replayed.
export async function createSession(
  userId: string,
  meta: SessionMetadata,
  existingFamilyId?: string,
): Promise<CreatedSession> {
  const familyId = existingFamilyId ?? randomUUID();
  const refreshToken = generateSecureToken(32);
  const session = await RefreshSession.create({
    userId,
    familyId,
    tokenHash: sha256(refreshToken),
    ip: meta.ip,
    userAgent: meta.userAgent,
    expiresAt: expiresIn(env.REFRESH_TOKEN_TTL),
  });
  return { sessionId: session._id.toString(), refreshToken };
}

export type RotateResult =
  | { status: 'ok'; sessionId: string; refreshToken: string; userId: string; roles: Role[] }
  | { status: 'invalid' }
  | { status: 'reuse' };

export async function rotateSession(
  presentedToken: string,
  meta: SessionMetadata,
): Promise<RotateResult> {
  // The opaque refresh token is the lookup key: we find the session by the
  // SHA-256 of the presented token. This is exactly why only hashes are stored.
  const session = await RefreshSession.findOne({ tokenHash: sha256(presentedToken) });
  const now = new Date();
  if (!session || session.expiresAt <= now) {
    return { status: 'invalid' };
  }

  // A presented token whose session is already revoked (e.g. it was rotated
  // away moments ago) means the token was almost certainly stolen and replayed.
  // Burn the entire family so every device in it is signed out.
  if (session.revokedAt !== null) {
    await revokeFamily(session.familyId, 'family_revoked');
    return { status: 'reuse' };
  }

  const user = await User.findById(session.userId).lean();
  if (!user || user.status !== 'active') {
    await revokeFamily(session.familyId, 'family_revoked');
    return { status: 'invalid' };
  }

  await RefreshSession.updateOne(
    { _id: session._id },
    { $set: { revokedAt: now, revokedReason: 'rotated', lastUsedAt: now } },
  );

  const created = await createSession(user._id.toString(), meta, session.familyId);
  return { status: 'ok', ...created, userId: user._id.toString(), roles: user.roles };
}

async function revokeFamily(familyId: string, reason: SessionRevocationReason): Promise<void> {
  await RefreshSession.updateMany(
    { familyId, revokedAt: null },
    { $set: { revokedAt: new Date(), revokedReason: reason } },
  );
}

export async function revokeSessionById(
  sessionId: string,
  userId: string,
  reason: SessionRevocationReason,
): Promise<boolean> {
  const result = await RefreshSession.updateOne(
    { _id: sessionId, userId, revokedAt: null },
    { $set: { revokedAt: new Date(), revokedReason: reason } },
  );
  return result.matchedCount > 0;
}

export async function revokeAllSessions(
  userId: string,
  reason: SessionRevocationReason,
  options: { exceptSessionId?: string } = {},
): Promise<void> {
  const filter: Record<string, unknown> = { userId, revokedAt: null };
  if (options.exceptSessionId) filter._id = { $ne: options.exceptSessionId };
  await RefreshSession.updateMany(filter, {
    $set: { revokedAt: new Date(), revokedReason: reason },
  });
}

export async function listSessionsForUser(
  userId: string,
  currentSessionId: string,
): Promise<SessionPublic[]> {
  const sessions = await RefreshSession.find({ userId }).sort({ createdAt: -1 }).limit(50);
  return sessions.map((session) => ({
    id: session._id.toString(),
    ip: session.ip,
    userAgent: session.userAgent,
    createdAt: session.createdAt.toISOString(),
    lastUsedAt: session.lastUsedAt?.toISOString() ?? null,
    expiresAt: session.expiresAt.toISOString(),
    active: session.revokedAt === null && session.expiresAt > new Date(),
    current: session._id.toString() === currentSessionId,
  }));
}

export { SESSION_REVOCATION_REASONS };

File: java/src/main/java/com/ecommerce/backend/modules/auth/SessionService.java
package com.ecommerce.backend.modules.auth;

import com.ecommerce.backend.common.Role;
import com.ecommerce.backend.common.util.SecureTokens;
import com.ecommerce.backend.common.util.TtlParser;
import com.ecommerce.backend.config.AppConfig;
import com.ecommerce.backend.modules.user.User;
import com.ecommerce.backend.modules.user.UserRepository;
import java.time.Instant;
import java.util.List;
import java.util.UUID;
import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;
import org.springframework.stereotype.Service;

import static org.springframework.data.mongodb.core.query.Criteria.where;
import static org.springframework.data.mongodb.core.query.Query.query;

/**
 * Refresh-token sessions (mirrors {@code session.service.ts}). Raw refresh
 * tokens are opaque random values, never JWTs: a token carries no identity
 * itself, it is just a key into a server-side session row — which is what makes
 * per-session revocation and reuse detection possible.
 */
@Service
public class SessionService {

    private final MongoTemplate mongo;
    private final AppConfig config;
    private final UserRepository userRepository;

    public SessionService(MongoTemplate mongo, AppConfig config, UserRepository userRepository) {
        this.mongo = mongo;
        this.config = config;
        this.userRepository = userRepository;
    }

    public record SessionMeta(String ip, String userAgent) {}

    public record CreatedSession(String sessionId, String refreshToken) {}

    public record RotateResult(
            String status, String sessionId, String refreshToken, String userId, List<Role> roles) {

        public static RotateResult ok(String sessionId, String refreshToken, String userId, List<Role> roles) {
            return new RotateResult("ok", sessionId, refreshToken, userId, roles);
        }

        public static RotateResult invalid() {
            return new RotateResult("invalid", null, null, null, null);
        }

        public static RotateResult reuse() {
            return new RotateResult("reuse", null, null, null, null);
        }
    }

    public record SessionPublic(
            String id,
            String ip,
            String userAgent,
            String createdAt,
            String lastUsedAt,
            String expiresAt,
            boolean active,
            boolean current) {}

    public CreatedSession createSession(String userId, SessionMeta meta, String existingFamilyId) {
        String familyId = existingFamilyId != null ? existingFamilyId : UUID.randomUUID().toString();
        String refreshToken = SecureTokens.generateSecureToken(32);
        RefreshSession session = new RefreshSession();
        session.userId = userId;
        session.familyId = familyId;
        session.tokenHash = SecureTokens.sha256(refreshToken);
        session.ip = meta.ip();
        session.userAgent = meta.userAgent();
        session.expiresAt = Instant.now().plus(TtlParser.parse(config.refreshTokenTtl()));
        RefreshSession saved = mongo.save(session);
        return new CreatedSession(saved.id, refreshToken);
    }

    public RotateResult rotateSession(String presentedToken, SessionMeta meta) {
        RefreshSession session =
                mongo.findOne(query(where("tokenHash").is(SecureTokens.sha256(presentedToken))), RefreshSession.class);
        Instant now = Instant.now();
        if (session == null || session.expiresAt == null || !session.expiresAt.isAfter(now)) {
            return RotateResult.invalid();
        }

        // A presented token whose session is already revoked (e.g. rotated away
        // moments ago) means it was almost certainly stolen and replayed. Burn
        // the entire family so every device in it is signed out.
        if (session.revokedAt != null) {
            revokeFamily(session.familyId, "family_revoked");
            return RotateResult.reuse();
        }

        User user = userRepository.findById(session.userId).orElse(null);
        if (user == null || !User.STATUS_ACTIVE.equals(user.status)) {
            revokeFamily(session.familyId, "family_revoked");
            return RotateResult.invalid();
        }

        mongo.updateFirst(
                query(where("_id").is(session.id)),
                new Update().set("revokedAt", now).set("revokedReason", "rotated").set("lastUsedAt", now),
                RefreshSession.class);

        CreatedSession created = createSession(user.id, meta, session.familyId);
        return RotateResult.ok(created.sessionId(), created.refreshToken(), user.id, user.roles);
    }

    public void revokeFamily(String familyId, String reason) {
        mongo.updateMulti(
                query(where("familyId").is(familyId).and("revokedAt").is(null)),
                new Update().set("revokedAt", Instant.now()).set("revokedReason", reason),
                RefreshSession.class);
    }

    public boolean revokeSessionById(String sessionId, String userId, String reason) {
        var result = mongo.updateFirst(
                query(where("_id").is(sessionId).and("userId").is(userId).and("revokedAt").is(null)),
                new Update().set("revokedAt", Instant.now()).set("revokedReason", reason),
                RefreshSession.class);
        return result.getMatchedCount() > 0;
    }

    public void revokeAllSessions(String userId, String reason, String exceptSessionId) {
        Query filter = query(where("userId").is(userId).and("revokedAt").is(null));
        if (exceptSessionId != null && !exceptSessionId.isBlank()) {
            filter.addCriteria(where("_id").ne(exceptSessionId));
        }
        mongo.updateMulti(
                filter,
                new Update().set("revokedAt", Instant.now()).set("revokedReason", reason),
                RefreshSession.class);
    }

    public List<SessionPublic> listSessionsForUser(String userId, String currentSessionId) {
        Instant now = Instant.now();
        Query query = query(where("userId").is(userId))
                .with(Sort.by(Sort.Direction.DESC, "createdAt"))
                .limit(50);
        return mongo.find(query, RefreshSession.class).stream()
                .map(s -> new SessionPublic(
                        s.id,
                        s.ip,
                        s.userAgent,
                        s.createdAt == null ? null : s.createdAt.toString(),
                        s.lastUsedAt == null ? null : s.lastUsedAt.toString(),
                        s.expiresAt == null ? null : s.expiresAt.toString(),
                        s.revokedAt == null && s.expiresAt != null && s.expiresAt.isAfter(now),
                        s.id != null && s.id.equals(currentSessionId)))
                .toList();
    }
}
